server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java

```diff
@@ -0,0 +1,1051 @@
+        }
+        final BackupProvider backupProvider = getBackupProvider(zoneId);
+        LOG.debug("Listing external backup offerings for the backup provider configured for zone ID " + zoneId);
+        return backupProvider.listBackupOfferings(zoneId);
+    }
+
+    @Override
+    @ActionEvent(eventType = EventTypes.EVENT_VM_BACKUP_IMPORT_OFFERING, eventDescription = "importing backup offering", async = true)
+    public BackupOffering importBackupOffering(final ImportBackupOfferingCmd cmd) {
+        validateForZone(cmd.getZoneId());
+        final BackupOffering existingOffering = backupOfferingDao.findByExternalId(cmd.getExternalId(), cmd.getZoneId());
+        if (existingOffering != null) {
+            throw new CloudRuntimeException("A backup offering with external ID " + cmd.getExternalId() + " already exists");
+        }
+        if (backupOfferingDao.findByName(cmd.getName(), cmd.getZoneId()) != null) {
+            throw new CloudRuntimeException("A backup offering with the same name already exists in this zone");
+        }
+
+        final BackupProvider provider = getBackupProvider(cmd.getZoneId());
+        if (!provider.isBackupOffering(cmd.getZoneId(), cmd.getExternalId())) {
+            throw new CloudRuntimeException("Backup offering '" + cmd.getExternalId() + "' does not exist on provider " + provider.getName() + " on zone " + cmd.getZoneId());
+        }
+
+        final BackupOfferingVO offering = new BackupOfferingVO(cmd.getZoneId(), cmd.getExternalId(), provider.getName(),
+                cmd.getName(), cmd.getDescription(), cmd.getUserDrivenBackups());
+
+        final BackupOfferingVO savedOffering = backupOfferingDao.persist(offering);
+        if (savedOffering == null) {
+            throw new CloudRuntimeException("Unable to create backup offering: " + cmd.getExternalId() + ", name: " + cmd.getName());
+        }
+        LOG.debug("Successfully created backup offering " + cmd.getName() + " mapped to backup provider offering " + cmd.getExternalId());
+        return savedOffering;
+    }
+
+    @Override
+    public Pair<List<BackupOffering>, Integer> listBackupOfferings(final ListBackupOfferingsCmd cmd) {
+        final Long offeringId = cmd.getOfferingId();
+        final Long zoneId = cmd.getZoneId();
+        final String keyword = cmd.getKeyword();
+
+        if (offeringId != null) {
+            BackupOfferingVO offering = backupOfferingDao.findById(offeringId);
+            if (offering == null) {
+                throw new CloudRuntimeException("Offering ID " + offeringId + " does not exist");
+            }
+            return new Pair<>(Collections.singletonList(offering), 1);
+        }
+
+        final Filter searchFilter = new Filter(BackupOfferingVO.class, "id", true, cmd.getStartIndex(), cmd.getPageSizeVal());
+        SearchBuilder<BackupOfferingVO> sb = backupOfferingDao.createSearchBuilder();
```

Comment:
Can this be moved to the backupOfferingDao instead?

Reply:
This is kind of complex as it does joins and filters, with support for pagination. I'll see how we can refactor this.